### PR TITLE
Refresh credentials once

### LINKED ISSUE #62: Concurrent refreshCredentials issue
`refreshCredentials ` in https://github.com/dart-lang/oauth2/blob/master/lib/src/client.dart#L148 has concurrent issue , assume we have an expired access token and call `client.read` twice , `refreshCredentials` refreshed access token twice and in many OAuth server implementation , if you get access token and refresh it again , first access token will revoked (You can have only one valid access token)

We need some lock mechanism like https://pub.dev/packages/synchronized

### LINKED ISSUE #333: Concurrent refreshCredentials issue
`refreshCredentials ` in https://github.com/dart-lang/oauth2/blob/master/lib/src/client.dart#L148 has concurrent issue , assume we have an expired access token and call `client.read` twice , `refreshCredentials` refreshed access token twice and in many OAuth server implementation , if you get access token and refresh it again , first access token will revoked (You can have only one valid access token)

We need some lock mechanism like https://pub.dev/packages/synchronized

### PR DESCRIPTION
**Pull request by Kavantix**
Fixes dart-lang/tools#333

### CODE REVIEW

**Comment by Kavantix** [comment]
@jonasfj could you take a look?

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# 1.6.1
```
Comment (jonasfj) [inline CHANGELOG.md:1]: Need to also update `pubspec.yaml`

File: lib/src/client.dart [hunk 2]
```diff
@@ -151,14 +154,26 @@ class Client extends http.BaseClient {
...
+        _credentials = await _refreshingFuture;
+        _onCredentialsRefreshed?.call(_credentials);
+      } finally {
+        _refreshingFuture = null;
+      }
+    } else {
+      await _refreshingFuture;
+    }
 
     return this;
```
Comment (jonasfj) [inline lib/src/client.dart:178]: Nit, but consider:
```dart
return _refreshingFuture ?? Future.microtask(() async {
  try {
    _credentials = await credentials.refresh(....);
    _onCredentialsRefreshed?.call(_credentials);
    return this;
  } finally {
    _refreshingFuture = null;
  }
});
```

That way all the errors gets the same stack trace. Not important, I'm pretty your solution will work just as fine :D

**Review by jonasfj — APPROVED** [review]
If you update the version in `pubspec.yaml` I'll publish this.

File: lib/src/client.dart (hunk 2 above)
Comment (Kavantix) [reply lib/src/client.dart:178]: Assuming you mean `??=` that’s also a nice solution.
But personally I would actually prefer being able to see the different stack traces.
Then you can track which call actually triggered the refresh, although indeed usually it should not be important.

Comment (jonasfj) [reply lib/src/client.dart:178]: > Then you can track which call actually triggered the refresh, 

Hmm, that's a good point :D

**Pull request merged**